### PR TITLE
MLX: version bump

### PR DESCRIPTION
**Pull request by dhiltgen**
https://github.com/ml-explore/mlx/compare/b6368984b8e02a3fb3ee7986846c0fb85e1fccf7...ec5f6ff7b1e0b314fc13f9ba8fb7e422ed65484d

### CODE REVIEW

**Review by jessegross — APPROVED** [review]

**Comment by dhiltgen** [comment]
Refreshed this morning - required a small change to depthwise_conv to align with the latest symbols.

**Comment by jessegross** [comment]
I think it might be worth picking up one further commit - cbb4059d8c6652d1c2cce98c348a8a689eb275b8

**Comment by dhiltgen** [comment]
Now contains 2 commits - first is a minimal bump with MLX-C patch, and the second plumbs global scale through the models in case we want to split that out as a follow up.  For existing library models, this is a no-op since they do not contain global scales yet.  I've done some initial Model Optimizer runs on a few models and the results are:

Model | GPU | Library prompt | ModelOpt global prompt | Delta | Library generate | ModelOpt global generate | Raw delta
-- | -- | -- | -- | -- | -- | -- | --
Qwen3.6 35B-A3B NVFP4 | M5 Max | 4,659.79 | 4,541.30 | -2.54% | 156.48 | 199.83 | +27.70%
Qwen3.6 35B-A3B NVFP4 | M4 Pro | 871.96 | 845.58 | -3.03% | 81.85 | 105.69 | +29.13%
Nemotron 3.5 Lightning 30B-A3B | M5 Max | 5,115.19 | 5,237.93 | +2.40% | 138.01 | 162.47 | +17.72%
Nemotron 3.5 Lightning 30B-A3B | M3 Ultra | 2,907.36 | 2,932.45 | +0.86% | 140.83 | 146.48 | +4.01%


This was just a quick pass without refining dtype promotions, so there is most likely further optimizations possible to improve performance.

File: x/mlxrunner/mlx/ops_extra.go [hunk 1]
```diff
@@ -90,6 +89,16 @@ func QuantizedMatmul(x, w, scales, biases *Array, transpose bool, groupSize, bit
 }
 
 func GatherQMM(x, w, scales *Array, biases, lhsIndices, rhsIndices *Array, transpose bool, groupSize, bits int, mode string, sortedIndices bool) *Array {
+	return gatherQMM(x, w, scales, biases, lhsIndices, rhsIndices, transpose, groupSize, bits, mode, nil, sortedIndices)
+}
+
+// GatherQMMWithGlobalScale applies one global scale per quantized weight bank.
+// A nil globalScale is equivalent to GatherQMM.
+func GatherQMMWithGlobalScale(x, w, scales *Array, biases, lhsIndices, rhsIndices *Array, transpose bool, groupSize, bits int, mode string, globalScale *Array, sortedIndices bool) *Array {
```
Comment (jessegross) [inline x/mlxrunner/mlx/ops_extra.go]: Minor, but can we add the nullable global-scale argument to the existing `mlx_gather_qmm` and `GatherQMM` functions and update their callers? MLX-C [added global scales directly to `mlx_quantize` and `mlx_dequantize`](https://github.com/ml-explore/mlx-c/commit/1370f59879c2d9ccc07c550db17042e27c4f39f6), and our `Dequantize` wrapper was extended the same way.

Callers without a global scale can pass an empty `mlx_array` in C and `nil` in Go. That keeps one function at each layer and follows the existing global-scale APIs.

File: x/models/nemotron_h/nemotron_h.go [hunk 2]
```diff
@@ -904,6 +965,7 @@ func loadLayer(linears model.LinearFactory, tensors map[string]*mlx.Array, cfg *
 			moe.UpWeightQ = up.Weight
 			moe.UpScales = up.Scales
 			moe.UpBiases = up.Biases
+			moe.UpGlobalScales = up.GlobalScales
```
Comment (jessegross) [inline x/models/nemotron_h/nemotron_h.go:1006]: Can we extend `UpGlobalScales` and `DownGlobalScales` when folding shared experts? We add weight banks without adding scales, so the first forward fails when routed experts have global scales. Shared experts without one need MLX’s identity scale.

File: cmake/local.cmake [hunk 3]
```diff
@@ -188,7 +199,9 @@ if(OLLAMA_MLX_BACKENDS)
             CONFIGURE_COMMAND ""
             BUILD_COMMAND ""
             INSTALL_COMMAND ""
-            USES_TERMINAL_DOWNLOAD TRUE)
+            PATCH_COMMAND ${OLLAMA_MLX_C_COMPAT_PATCH_COMMAND}
```
Comment (jessegross) [inline cmake/local.cmake:205]: Can we apply this patch for local source overrides too? Both override paths skip it, so a clean checkout at the pinned MLX-C revision fails to build.

File: x/mlxrunner/mlx/ops_extra.go (hunk 1 above)
Comment (jessegross) [inline x/mlxrunner/mlx/ops_extra.go]: Can we extend `mlx_gather_qmm` and `GatherQMM` instead of adding `WithGlobalScale` versions? That follows the existing pattern in MLX-C and our `Dequantize` wrapper.

File: x/mlxrunner/mlx/ops_extra.go [hunk 4]
```diff
@@ -105,9 +114,12 @@ func GatherQMM(x, w, scales *Array, biases, lhsIndices, rhsIndices *Array, trans
 	if rhsIndices != nil {
 		rhs = rhsIndices.ctx
 	}
-
 	out := New("GATHER_QMM")
-	mlxCheck(C.mlx_gather_qmm(&out.ctx, x.ctx, w.ctx, scales.ctx, b, lhs, rhs, C.bool(transpose), optGroupSize, optBits, cMode, C.bool(sortedIndices), DefaultStream().ctx))
+	if globalScale != nil {
```
Comment (jessegross) [inline x/mlxrunner/mlx/ops_extra.go]: Can we handle CUDA’s global scale by multiplying the gathered output in `GatherQMM`? That keeps the experts quantized and avoids the memory and weight traffic of dequantizing the whole bank. Metal can keep the native path. Both should accept MLX’s scale convention, with backend differences handled inside the wrapper.

Let’s move the existing NVFP4 multiply-and-cast from `QuantizedLinear` into `QuantizedMatmul` too.

We should compare logits against the dequantized reference to check rounding and nonfinite outputs.

File: x/mlxrunner/model/quant.go [hunk 5]
```diff
@@ -5,6 +5,32 @@ import (
...
 
+// nvfp4MaxProduct is the product of the maximum E4M3 and E2M1 values.
+const nvfp4MaxProduct = 448 * 6
+
+// PrepareGatherQMMGlobalScale converts a ModelOpt NVFP4 dequantization
+// multiplier to the amax convention that MLX's gather_qmm expects. MLX requires
+// one global scale per expert, so a checkpoint-wide scalar is expanded to the
+// expert count. The native global-scale path is currently available only on
+// Metal.
+func PrepareGatherQMMGlobalScale(globalScale *mlx.Array, mode string, numExperts int) (*mlx.Array, bool) {
```
Comment (jessegross) [inline x/mlxrunner/model/quant.go]: With scaling handled in the wrappers, this should just prepare and return the scale: handle nil, convert to float32, broadcast to the expert count, and multiply by `448 * 6`.

Can we drop the mode/backend checks, support boolean, and associated dequantization fallback? Broadcasting stays here, using `BroadcastTo` instead of zeros plus add.

File: x/models/qwen3_5/qwen3_5.go [hunk 6]
```diff
@@ -506,6 +518,7 @@ func fuseGateUpProjections(gate, up *stackedExpertWeights) *stackedExpertWeights
 		return nil
 	}
 	if gate.Scales != nil && up.Scales != nil &&
+		gate.GlobalScales == nil && up.GlobalScales == nil &&
```
Comment (jessegross) [inline x/models/qwen3_5/qwen3_5.go]: Can we also fuse gate/up when their quantization parameters match and both use the same global scale for each expert? Carry that scale into the fused projection, and keep the separate path when the quantization or scales differ. This applies to Laguna’s routed experts too.

File: x/mlxrunner/mlx/ops_extra.go [hunk 7]
```diff
@@ -105,12 +133,35 @@ func GatherQMM(x, w, scales *Array, biases, lhsIndices, rhsIndices *Array, trans
...
+		out = mulGatherQMMGlobalScale(out, globalScale, rhsIndices)
+	}
 	return out
 }
 
+// mulGatherQMMGlobalScale scales gathered output rows by the per-expert
+// global scale.
+func mulGatherQMMGlobalScale(out, globalScale, rhsIndices *Array) *Array {
+	perExpert := DivScalar(globalScale, Nvfp4MaxProduct)
+	scaleRows := Take(perExpert, rhsIndices, 0)
```
Comment (jessegross) [inline x/mlxrunner/mlx/ops_extra.go:156]: Flatten the global-scale bank before indexing it in the CUDA wrapper. MLX’s RHS indices address the flattened weight batch dimensions, so taking axis 0 gives incorrect results for banks with multiple batch axes.

File: x/models/gemma4/gemma4.go [hunk 8]
```diff
@@ -218,42 +222,112 @@ func collectExpertProjection(tensors map[string]*mlx.Array, cfg *TextConfig, pre
...
+		Scales: mlx.Stack(scales, 0).Clone(),
+		Bits:   parts[0].bits, GroupSize: parts[0].groupSize, Mode: parts[0].mode,
 	}
-	if len(biases) == len(weights) {
+	if hasBiases {
 		out.Biases = mlx.Stack(biases, 0).Clone()
 	}
+	if hasGlobalScales {
+		out.GlobalScales = mlx.Reshape(mlx.Stack(globalScales, 0).Clone(), int32(len(parts)))
+		out.DequantGlobalScales = mlx.Reshape(mlx.Stack(dequantGlobalScales, 0).Clone(), int32(len(parts)))
```
Comment (jessegross) [inline x/models/gemma4/gemma4.go]: Normalize each raw global scale before stacking. Scalar shapes `[]` and `[1]` are both valid, but mixing them currently panics here. This also applies to Cohere and GLM.

File: x/models/qwen3_5/qwen3_5_test.go [hunk 9]
```diff
@@ -121,6 +122,129 @@ func TestSupportsGatherQMM(t *testing.T) {
...
+			QuantBits:        4,
+			QuantMode:        "nvfp4",
+			NumExpertsPerTok: 2,
+		}
+		prefix := "model.layers.0"
+
+		// Nonzero nvfp4 payloads: packed fp4 codes cycle through all 16 values
+		// and every group scale is a power of two, so the dequantized weights
+		// are exact and the dense path is a faithful reference.
+		packed := make([]uint32, width*width/8)
```
Comment (jessegross) [inline x/models/qwen3_5/qwen3_5_test.go]: Include `experts` in both buffer allocations. The tensor shapes already include that dimension, so `FromValues` reads past both buffers and the test can pass using unrelated memory.

File: x/mlxrunner/mlx/ops_extra.go [hunk 10]
```diff
@@ -61,19 +81,19 @@ func Dequantize(w, scales, biases *Array, groupSize, bits int, mode string, glob
...
+			gs = ExpandDims(gs, -1)
 		}
 		outType := out.DType()
 		out = Mul(out, gs).AsType(outType)
 	}
 	return out
 }
 
-func QuantizedMatmul(x, w, scales, biases *Array, transpose bool, groupSize, bits int, mode string) *Array {
+// QuantizedMatmul takes the checkpoint's globalScale; MLX's quantized_matmul
```
Comment (jessegross) [inline x/mlxrunner/mlx/ops_extra.go]: Use MLX’s `globalScale` representation consistently in `GatherQMM`, `Dequantize`, and `QuantizedMatmul`. Convert checkpoint scales during loading and divide by `Nvfp4MaxProduct` in wrappers that apply the multiplier. The missing-scale defaults in Qwen’s `concatGlobalScales` and DFlash’s `perRowGlobal` also need to use `Nvfp4MaxProduct`. Update Laguna’s direct scale uses so it can share `GlobalScales` and drop `DequantGlobalScales`.

File: x/models/qwen3_5/qwen3_5.go [hunk 11]
```diff
@@ -592,25 +604,39 @@ func collectPerExpertProjection(tensors map[string]*mlx.Array, cfg *Config, useQ
 	weights := make([]*mlx.Array, 0, numExperts)
 	scales := make([]*mlx.Array, 0, numExperts)
 	biases := make([]*mlx.Array, 0, numExperts)
+	globalScales := make([]*mlx.Array, 0, numExperts)
+	dequantGlobalScales := make([]*mlx.Array, 0, numExperts)
```
Comment (jessegross) [inline x/models/qwen3_5/qwen3_5.go]: Replace `dequantGlobalScales` with a check that global scales are present for all experts or none. Only the slice’s length is read. Dequantization already uses each expert’s `globalScale` directly inside the loop.

**Review by jessegross — COMMENTED** [review]
Good to merge once the comments are addressed.
